types/src/transaction/transaction_target.rs

```diff
@@ -23,6 +21,144 @@ use rand::{Rng, RngCore};
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
 
+const VM_CASPER_V1_TAG: u8 = 0;
+const VM_CASPER_V2_TAG: u8 = 1;
+const TRANSFERRED_VALUE_INDEX: u16 = 1;
+const SEED_VALUE_INDEX: u16 = 2;
+
+#[derive(Clone, Ord, PartialOrd, Eq, PartialEq, Hash, Serialize, Deserialize, Debug)]
+#[cfg_attr(feature = "datasize", derive(DataSize))]
+#[cfg_attr(
+    feature = "json-schema",
+    derive(JsonSchema),
+    schemars(description = "Session params of a TransactionTarget.")
+)]
+#[serde(deny_unknown_fields)]
+pub enum TransactionRuntimeParams {
+    VmCasperV1,
+    VmCasperV2 {
+        /// The amount of motes to transfer before code is executed.
+        ///
+        /// This is for protection against phishing attack where a malicious session code drains
+        /// the balance of the caller account. The amount stated here is the maximum amount
+        /// that can be transferred from the caller account to the session account.
+        transferred_value: u64,
+        /// The seed for the session code that is used for an installer.
+        seed: Option<[u8; 32]>,
+    },
+}
+
+impl TransactionRuntimeParams {
+    /// Returns the contract runtime tag.
+    pub fn contract_runtime_tag(&self) -> ContractRuntimeTag {
+        match self {
+            TransactionRuntimeParams::VmCasperV1 => ContractRuntimeTag::VmCasperV1,
+            TransactionRuntimeParams::VmCasperV2 { .. } => ContractRuntimeTag::VmCasperV2,
+        }
+    }
+
+    pub fn seed(&self) -> Option<[u8; 32]> {
+        match self {
+            TransactionRuntimeParams::VmCasperV1 => None,
+            TransactionRuntimeParams::VmCasperV2 { seed, .. } => *seed,
+        }
+    }
+}
+
+impl ToBytes for TransactionRuntimeParams {
+    fn to_bytes(&self) -> Result<Vec<u8>, Error> {
+        match self {
+            TransactionRuntimeParams::VmCasperV1 => {
+                CalltableSerializationEnvelopeBuilder::new(vec![
+                    crate::bytesrepr::U8_SERIALIZED_LENGTH,
```

Comment:
I personally prefer to put the vec of fields in a separate function and call it both in to_bytes and serialized_length. That way - if any new fields are introduced - only one place needs updating. You can see that in TransactionV1#serialized_field_lengths

Reply:
Addressed